### PR TITLE
Improve Python bindings after R/U/G types

### LINKED ISSUE #747: Implement Python bindings for Mulliken-partitioning
Since we want to be able to do our research in a Jupyter Notebook, we need Python bindings for Mulliken-partitioning.
- [x] Restricted case
- [x] Unrestricted case

### LINKED ISSUE #784: Python bindings can no longer quantize operators
**Describe the bug**
The new compile-time correct operators don't have any python bindings yet. This causes a type error while trying to run several calculations, like Hartree-Fock.

**To Reproduce**
Screenshot 2020-11-12 at 10 42 16

**Expected behavior**
Quantizing the operator returns a new type (`RSQOneElectronOperator` in this case). B ut this new type is not yet recognised by python.

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**
This PR aims to (re-)enable Python bindings for Mulliken calculations: #747.

**Related issues**
- As a side effect of these implementations, this PR closes #784;
- It is one of the parts related to #762;
- It closes #747.

**Don't forget:**
- [ ] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

**Comment by lelemmen** [comment]
@xdvriend I've requested your review for this PR.

I went rather quickly in rewriting the current Python bindings, so I might have forgotten a couple of things. (The total rewrite is not done yet, this will be done in a second PR in order to resolve #762)

I focused mainly on being able to do unrestricted Mulliken calculations; you can find an updated Jupyter Notebook under `gqcpy/examples/Mulliken population.ipynb`, showcasing how to do the revamped Mulliken partitioning with the new APIs.

File: gqcpy/src/DensityMatrix/SpinResolved1DM_bindings.cpp [hunk 1]
```diff
@@ -16,62 +16,75 @@
...
+    // Expose the `SpinResolvedBase` API to the Python class.
+    bindSpinResolvedBaseInterface(py_SpinResolved1DM_d);
+
+
+    // Expose the `BasisTransformable` API to the Python class.
+    bindBasisTransformableInterface(py_SpinResolved1DM_d);
 }
 
+
 }  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/DensityMatrix/SpinResolved1DM_bindings.cpp:90]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/DensityMatrix/SpinResolved2DM_bindings.cpp [hunk 2]
```diff
@@ -27,52 +28,42 @@ namespace py = pybind11;
...
+            &SpinResolved2DM<double>::orbitalDensity,
+            "The orbital (total, spin-summed) two-electron density matrix.");
+
+
+    // Expose the `DoublySpinResolvedBase` API to the Python class.
+    bindDoublySpinResolvedBaseInterface(py_SpinResolved2DM_d);
 }
 
+
 }  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/DensityMatrix/SpinResolved2DM_bindings.cpp:69]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/Operator/SecondQuantized/RSQOneElectronOperator_bindings.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    bindSimpleSQOneElectronOperatorInterface(py_VectorRSQOneElectronOperator_d);
+
+
+    py::class_<VectorRSQOneElectronOperator<complex>> py_VectorRSQOneElectronOperator_cd {module, "VectorRSQOneElectronOperator_cd", "A (complex) restricted one-electron operator, which is suited for expressing non-relativistic (spin-free) one-electron operators."};
+
+    bindSimpleSQOneElectronOperatorInterface(py_VectorRSQOneElectronOperator_cd);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/Operator/SecondQuantized/RSQOneElectronOperator_bindings.cpp:64]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/Operator/SecondQuantized/USQOneElectronOperatorComponent_bindings.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    bindSimpleSQOneElectronOperatorInterface(py_VectorUSQOneElectronOperatorComponent_d);
+
+
+    py::class_<VectorUSQOneElectronOperatorComponent<complex>> py_VectorUSQOneElectronOperatorComponent_cd {module, "VectorUSQOneElectronOperatorComponent_cd", "One of the spin components of a (complex) unrestricted one-electron operator, i.e. either the alpha or beta part."};
+
+    bindSimpleSQOneElectronOperatorInterface(py_VectorUSQOneElectronOperatorComponent_cd);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/Operator/SecondQuantized/USQOneElectronOperatorComponent_bindings.cpp:64]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/Operator/SecondQuantized/USQTwoElectronOperator_bindings.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Define the Python class for `USQTwoElectronOperator`.
+    py::class_<ScalarUSQTwoElectronOperator<double>> py_ScalarUSQTwoElectronOperator_d {module, "ScalarUSQTwoElectronOperator_d", "A class that represents a (real) 'unrestricted second-quantized two-electron operator' suitable for the projection of the non-relativistic Hamiltonian onto an unrestricted spinor basis. It holds the tensor representation of its parameters for both spin components and both mixed spin components, which are (usually) integrals over first-quantized operators."};
+
+    // Expose the `DoublySpinResolvedBase` API to the Python class.
+    bindDoublySpinResolvedBaseInterface(py_ScalarUSQTwoElectronOperator_d);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/Operator/SecondQuantized/USQTwoElectronOperator_bindings.cpp:51]:
```suggestion
}  // namespace gqcpy

```

**Review by xdvriend — APPROVED** [review]
- The binding of the interfaces makes the overall implementation of the other python bindings so much clearer, nice!
- Wow, that unrestricted Mulliken example looks super clean!

Awesome PR! Other than the few files that had no new line at the end (and a few typo's I discovered later while using the new bindings), no remarks!

File: gqcpy/examples/Mulliken population.ipynb [hunk 6]
```diff
@@ -59,58 +66,243 @@
...
+   "outputs": [],
+   "source": [
+    "molecule = gqcpy.Molecule.ReadXYZ(\"../../gqcp/tests/data/CO_mulliken.xyz\" , +1)  # '+1': Create CO+.\n",
+    "N_alpha = molecule.numberOfElectronPairs() + 1\n",
+    "N_beta = molecule.numberOfElectronPairs()\n",
+    "\n",
+    "u_spin_orbital_basis = gqcpy.USpinOrbitalBasis_d(molecule, \"STO-3G\")\n",
+    "S_unrestricted = u_spin_orbital_basis.quantizeOverlapOperator()\n",
+    "\n",
+    "hamiltonian = gqcpy.RSQHamiltonian.Molecular(spin_orbital_basis, molecule)  # We UHF calculations with the Hamiltonian in the basis of the restricted atomic spin-orbitals. "
```
Comment (xdvriend) [inline gqcpy/examples/Mulliken population.ipynb:209]:
```suggestion
    "hamiltonian = gqcpy.RSQHamiltonian.Molecular(spin_orbital_basis, molecule)  # We perform UHF calculations with the Hamiltonian in the basis of the restricted atomic spin-orbitals. "
```

File: gqcpy/src/Basis/MullikenPartitioning/UMullikenPartitioning_bindings.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+void bindUMullikenPartitioning(py::module& module) {
+
+    // Define the Python class for `UMullikenPartitioning`.
+    py::class_<UMullikenPartitioning<double>> py_UMullikenPartitioning_d {module, "UMullikenPartitioning_d", "An unrestricted Mulliken-based partitioning of an AO basis."};
+
+
+    // Expose the "Mulliken partitioning" interface to the Python class.
+    bindMullikenPartitioningMatricesInterface(py_UMullikenPartitioning_d);
+
+    // Exponse the `SpinResolvedBase` interface to the Python class.
```
Comment (xdvriend) [inline gqcpy/src/Basis/MullikenPartitioning/UMullikenPartitioning_bindings.cpp:46]:
```suggestion
    // Expose the `SpinResolvedBase` interface to the Python class.
```

File: gqcpy/include/interfaces.hpp [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+    bindBasisTransformableInterface(py_class);
+}
+
+
+/**
+ *  Add Python bindings for the parameter access for scalar two-electron operators.
+ * 
+ *  @tparam Class               The type of the Pybind11 `class_` (generated by the compiler).
+ *  
+ *  @param py_class             The Pybind11 `class_` that should obtain acess bindings like a scalar two-electron operator.
```
Comment (xdvriend) [inline gqcpy/include/interfaces.hpp:664]:
```suggestion
 *  @param py_class             The Pybind11 `class_` that should obtain access bindings like a scalar two-electron operator.
```

File: gqcpy/src/Operator/SecondQuantized/SQHamiltonian_bindings.cpp [hunk 9]
```diff
@@ -15,58 +15,122 @@
...
+
+        .def(ScalarSQOneElectronOperator() + py::self)
+
+        .def(py::self -= ScalarSQOneElectronOperator())
+
+        .def(py::self - ScalarSQOneElectronOperator())
+
+
+        /*
+         *  MARK: Operations related to one-electron operators
```
Comment (xdvriend) [inline gqcpy/src/Operator/SecondQuantized/SQHamiltonian_bindings.cpp:107]:
```suggestion
         *  MARK: Operations related to two-electron operators
```

File: gqcpy/src/QCMethod/CI/CIEnvironment_bindings.cpp [hunk 10]
```diff
@@ -1,73 +1,73 @@
...
+        py::arg("V"),
+        documentation.c_str());
+}
+
+
+void bindCIEnvironments(py::module& module) {
+
+    auto submodule = module.def_submodule("CIEnvironment");
+
+    bindCIEnvironment<RSQHamiltonian<double>, SpinResolvedONVBasis>(submodule, "Return an environment suitable for solving spin-resolved FCI eigenvalue problems.");
```
Comment (xdvriend) [inline gqcpy/src/QCMethod/CI/CIEnvironment_bindings.cpp:69]:
```suggestion
    bindCIEnvironment<RSQHamiltonian<double>, SpinResolvedONVBasis>(submodule, "Return an environment suitable for solving spin-resolved FCI eigenvalue problems.");
    bindCIEnvironment<USQHamiltonian<double>, SpinResolvedONVBasis>(submodule, "Return an environment suitable for solving spin-resolved FCI eigenvalue problems.");
```

In order to perform the spin-constrained CI calculations, I need the ability to work with a `USQHamiltonian` in my CI environment. I have added this here as a suggestion. If this was intentionally done differently, just say so and feel free to ignore my suggestion here.

Comment (lelemmen) [reply gqcpy/src/QCMethod/CI/CIEnvironment_bindings.cpp:69]: Thanks! I missed this extra binding for unrestricted CI to work, indeed!

**Comment by lelemmen** [comment]
@xdvriend I've added an example for unrestricted CI in the `FCI.ipynb` notebook, to confirm that these kinds of calculations are now enabled!

**Pull request merged**